cps/src/main/java/org/jenkinsci/plugins/workflow/cps/replay/ReplayFlowFactoryAction.java

```diff
@@ -0,0 +1,103 @@
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import javax.annotation.Nonnull;
+import org.jenkinsci.plugins.workflow.cps.CpsFlowExecution;
+import org.jenkinsci.plugins.workflow.cps.CpsFlowFactoryAction2;
+import org.jenkinsci.plugins.workflow.cps.steps.LoadStepExecution;
+import org.jenkinsci.plugins.workflow.flow.FlowDefinition;
+import org.jenkinsci.plugins.workflow.flow.FlowExecutionOwner;
+
+/**
+ * Attached to a run that is a replay of an earlier one.
+ */
+class ReplayFlowFactoryAction extends InvisibleAction implements CpsFlowFactoryAction2, Queue.QueueAction {
+
+    private static final Logger LOGGER = Logger.getLogger(ReplayFlowFactoryAction.class.getName());
+
+    private String mainScript;
+    private final Map<String,String> otherScripts;
+    private transient final boolean sandbox;
+    
+    ReplayFlowFactoryAction(@Nonnull String mainScript, @Nonnull Map<String,String> otherScripts, boolean sandbox) {
+        this.mainScript = mainScript;
+        this.otherScripts = new HashMap<String,String>(otherScripts);
+        this.sandbox = sandbox;
+    }
+
+    @Override public CpsFlowExecution create(FlowDefinition def, FlowExecutionOwner owner, List<? extends Action> actions) throws IOException {
+        String script = mainScript;
+        mainScript = null; // minimize build.xml size
+        return new CpsFlowExecution(script, sandbox, owner);
+    }
+
+    @Override public boolean shouldSchedule(List<Action> actions) {
+        return true; // do not coalesce
+    }
+
+    @Extension public static class ReplacerImpl implements LoadStepExecution.Replacer {
+
+        @Override public String replace(String text, CpsFlowExecution execution, String clazz, TaskListener listener) {
+            try {
+                Queue.Executable executable = execution.getOwner().getExecutable();
+                if (executable instanceof Run) {
+                    ReplayFlowFactoryAction action = ((Run) executable).getAction(ReplayFlowFactoryAction.class);
+                    if (action != null) {
+                        String newText = action.otherScripts.remove(clazz);
+                        if (newText != null) {
+                            listener.getLogger().println("Replacing Groovy text with edited version");
+                            return newText;
+                        } else {
+                            listener.getLogger().println("Warning: no replacement Groovy text found for " + clazz);
+                        }
+                    } else {
+                        LOGGER.log(Level.FINE, "{0} was not a replay", executable);
+                    }
+                } else {
+                    LOGGER.log(Level.FINE, "{0} was not a run at all", executable);
+                }
+            } catch (IOException x) {
+                LOGGER.log(Level.WARNING, null, x);
+            }
+            return text;
```

Comment:
So what is the `text` arg used for? I see (later) that it's "the original text", which seems strange.